README
Nascent Order for Final Fantasy VI
a supplemental randomizer by emberling
version 0.2.0

The main goal of this project is to build on top of FF6 Beyond Chaos to create
the randomized experience that most appeals to me. I'm fascinated by randomly
generated art pieces and story projects, things like @MagicRealismBot and
botnik.org. I'm not looking to challenge myself (gameplay wise) or to spark
hilarious moments on twitch dot teevee so much as I'm looking to create
fresh experiences that would be believable as alternate-universe versions
of FF6.

NaO is designed specifically for use with a ROM already run through Beyond
Chaos, but it is also designed to be highly flexible and to work with any
hacked ROM you give it! Some may require some reconfiguration, though, and
there are many permutations I haven't yet coded in. But in general, while
it's not the main intent of this program, it CAN be used as a tool to
insert and repack music for your non-random hack, for instance.

My rewrites and mods, and the modules contained herein, are
generally built around the idea of *curated* randomness. Fire can be
red, or yellow, or orange, or even blue, but not green. Grandpa can be
a knight or a ninja or a mime, but not a child.

Or you can turn that off, if you'd rather. I'm not picky about what YOU do.

Many of the Beyond Chaos systems are based around taking existing FF6 systems
-- palettes, spells, items, enemy AI -- in new, surprising, and incongruous
directions. My reimaginings generally flip this around, building those things from
scratch while trying to avoid incongruity as much as possible, relying on the
inevitable ineptitude of machine generation to produce those surprising
moments. BC's methods work great for streaming and community, and for
people always seeking new challenges. But I like different things than that,
personally, so here we are.

So far, NaO does these things:
    - Music randomizer
        - imports external music from MML files (included)
        - shuffles music appropriately (or inappropriately, if desired)
        - adds four additional slots for battle songs
            by default, this creates a different battle theme for WoB
            (outside and inside), WoR (outside and inside), and event battles.
        - Largely preserves music-format SFX used in WoR, train, zozo
        - Final boss music randomizer: shuffles and replaces the three
            individual parts of the first Dancing Mad track (three tiers)
    - Instrument sample expansion
        Adds a variety of new instruments for enhanced imported music.
    - Character palette rework
        Should be both more vivid and more consistent/less glitchy than BC.
        - Includes skintone randomization from a curated list of possibilities
        - Some default sprites require minor adjustments to work well with this
        mode, these are automatically applied.
        - Consistency maintained for some palettes that should logically match
    - Pixel portrait mode
        To solve the problem of custom sprites lacking functional portraits,
        this automatically generates serviceable portraits from the sprite
        itself. By default replaces only portraits that appear to be
        placeholders; configurable to replace all portraits.
    - Field sprite randomizer
        A slightly more advanced version of BC's 'partyparty makeover'.
        All playable characters and all NPCs with walking animations are
        randomized from a pool of custom and original sprites. Constraints
        are used so that the new set of sprites generally matches the story
        and function of the original.
        Character names are also randomized.
        CtrlxZ's v1.5 sprite pack is included, plus a few extra.
    - misc. tweaks
        Optionally change BC's "original character name" field to a class name.
        
On the horizon / work currently in progress:
    - Enemy sprite randomizer
        Including custom sprites. Various enemy properties and effects should
        be inferrable from the chosen graphics, as they are in a non-randomized
        game: e.g., flying.
    - Field sprite randomizer improvements
        Decoupling NPCs that use the same sprite with different palettes

Future plans:
    - Additional song slots for e.g. a true "hurry!" or "dungeon" track
    - Adjust script to reflect randomizations, e.g. pronouns
    - Optional balance tweaks, aimed at eliminating the recurring BC mechanic
        of things that seem cool but have a small but inevitable chance to
        flat out kill you. 
    - Items, and spell effects
    - Formula and ability adjustments to better support above

CHANGELOG
0.2.0:
    -new mode: 'c' -- randomize character names, sprites, and portraits
        This supersedes Beyond Chaos' 'partyparty makeover' mode.
        All field sprites with walking animations are randomized.
        Sprites without portraits use ghost if mode 'x' is not selected.
    -new mode: 'i' -- imports a custom instrument set, greatly expanding
        the versatility and range of custom music
    -changed mode: 'b' -- this now assigns music based on location rather
        than enemy level, with earlier locations receiving higher priority.
        The location breakdown is: World of Balance overworld encounters,
        World of Balance dungeon encounters, all event battles, World of
        Ruin overworld encounters, World of Ruin dungeon encounters.
        - The keyword 'ONEWINGEDLEAFER' engages the old behavior.
        - Also fixed a bug that was resulting in incorrect song assignment.
    -'n' -- class names are now loaded from custom/SpriteImports.txt instead
        of hardcoded
    -'m' -- now randomizes individual sections of tier boss music
    -'m' -- now loads MML files via mfvitools instead of raw binary data
    
0.1.0:
    -new mode: 'x' -- generates portraits based on sprites. this draws from the top 16x16 of the battle ready pose, or bottom 16x16 if the top 8 pixel rows of the pose are empty. by default, only portraits which fulfill one of these criteria will be replaced:
        - is a portrait from the base game, but the sprite it is applied to is not that character's original sprite, nor a known variation
        - is a duplicate of another character's portrait
        The keyword 'PIXELPARTY' ignores these constraints and applies consistent faux-retro portraits to every character.
    -new mode: 'p' -- generates random palettes for characters, actors, and NPCs. this uses a vastly different and (in my opinion) improved method compared to Beyond Chaos, greatly improving its functionality with custom sprites and improving the flexibility available to custom sprite makers, who previously had to avoid the pitfalls of the dreaded red/green clown palette used by Kefka and the zombie-like pale monstrosity used by Umaro.
        - Biased toward variety -- palettes will now be used a roughly equal number of times throughout your party, every time. No more seeds where everyone looks the same at first glance.
        - Biased toward human perception -- instead of treating the entire color spectrum equally, this mode skews more toward the areas of the spectrum where color differentiation is most apparent.
        - A much larger variety of skin color than the original game.
        - Less immersion-breaking, in service of my eventual goal to have a randomly generated FF6 that is every bit as internally consistent and polished as the original
            - Traitorous Imperial generals now maintain a consistent palette even when imprisoned.
            - Twins, including royal ones, now share the same palette, and thus generally the same hair and skin color.
        - Consistent pattern for sprite artists to work from:
            0. transparency
            1. borders (these change color during various status effects)
            2. white -- this is now always lightly tinted toward HairColor
            3. near-black -- this is also always tinted toward HairColor
            4-5. hair -- this is the lightest of the three main color types, with almost always relatively low saturation.
            6-7. skin -- rather than being generated, this is chosen from a curated list in sprites.cfg
            8-9. clothing -- this is the darkest of the three color types, with a wide variety of saturation. Clothing Light and Hair Dark are generally around the same brightness. Clothing Dark is between near-black and Hair Dark, and can be used in a gradient with those with reasonable effectiveness. (2-4-5-9-3 light-to-dark)
            10-11. accessory -- this has variable brightness and usually high saturation, though it has a chance for very low saturation instead (gray).
            12-15. town palette -- these are used only on NPCs and still somewhat of a work in progress. their parameters are similar to accessory, though 12-13 is a bit darker and 14-15 less saturated. these are in dark-light order instead of light-dark, which matches the original usage EXCEPT for palette 5, which was used only by Ultros and Chupon.

mmltbl.py

note_tbl = {
    "c": 0x0,
    "d": 0x2,
    "e": 0x4,
    "f": 0x5,
    "g": 0x7,
    "a": 0x9,
    "b": 0xB,
    "^": 0xC,
    "r": 0xD }
    
length_tbl = { 
    1   : (0, 0xC0),
    2   : (1, 0x60),
    3   : (2, 0x40),
    "4.": (3, 0x48),
    4   : (4, 0x30),
    6   : (5, 0x20),
    "8.": (6, 0x24),
    8   : (7, 0x18),
    12  : (8, 0x10),
    16  : (9, 0x0C),
    24  : (10, 0x08),
    32  : (11, 0x06),
    48  : (12, 0x04),
    64  : (13, 0x03) }

r_length_tbl = {
    0: "1",
    1: "2",
    2: "3",
    3: "4.",
    4: "4",
    5: "6",
    6: "8.",
    7: "8",
    8: "12",
    9: "", #default
    10: "24",
    11: "32",
    12: "48",
    13: "64" }

command_tbl = {
    ("@", 1) : 0xDC, #program
    ("|", 1) : 0xDC, #program (hex param)
    ("%a", 0): 0xE1, #reset ADSR
    ("%a", 1): 0xDD, #set attack
    ("%b", 1): 0xF7, #echo feedback (rs3)
    ("%b", 2): 0xF7, #echo feedback (ff6)
    ("%c", 1): 0xCF, #noise clock
    ("%d0", 0): 0xFC,#drum mode off (rs3)
    ("%d1", 0): 0xFB,#drum mode on (rs3)
    ("%e0", 0): 0xD5,#disable echo
    ("%e1", 0): 0xD4,#enable echo
    ("%f", 1): 0xF8, #filter (rs3)
    ("%f", 2): 0xF8, #filter (ff6)
    ("%g0", 0): 0xE7,#disable roll (enable gaps between notes)
    ("%g1", 0): 0xE6,#enable roll (disable gaps between notes)
    ("%k", 1): 0xD9, #set transpose
    ("%l0", 0): 0xE5,#disable legato 
    ("%l1", 0): 0xE4,#enable legato
    ("%n0", 0): 0xD1,#disable noise
    ("%n1", 0): 0xD0,#enable noise
    ("%p0", 0): 0xD3,#disable pitch mod
    ("%p1", 0): 0xD2,#enable pitch mod
    ("%r", 0): 0xE1, #reset ADSR
    ("%r", 1): 0xE0, #set release
    ("%s", 0): 0xE1, #reset ADSR
    ("%s", 1): 0xDF, #set sustain
    ("%v", 1): 0xF2, #set echo volume
    ("%v", 2): 0xF3, #echo volume envelope
    ("%x", 1): 0xF4, #set master volume
    ("%y", 0): 0xE1, #reset ADSR
    ("%y", 1): 0xDE, #set decay
    #("j", 1): 0xF5 - jump out of loop after n iterations
    #("j", 2): 0xF5 - jump to marker after n iterations
    ("k", 1): 0xDB,  #set detune
    ("m", 0): 0xCA,  #disable vibrato
    ("m", 1): 0xDA,  #add to transpose
    ("m", 2): 0xC8,  #pitch envelope (portamento)
    ("m", 3): 0xC9,  #enable vibrato
    ("o", 1): 0xD6,  #set octave
    ("p", 0): 0xCE,  #disable pan sweep
    ("p", 1): 0xC6,  #set pan
    ("p", 2): 0xC7,  #pan envelope
    ("p", 3): 0xCD,  #pansweep
    ("s0", 1): 0xE9, #play sound effect with voice A
    ("s1", 1): 0xEA, #play sound effect with voice B
    ("t", 1): 0xF0,  #set tempo
    ("t", 2): 0xF1,  #tempo envelope
    ("u0", 0): 0xFA, #clear output code
    ("u1", 0): 0xF9, #increment output code
    ("v", 0): 0xCC,  #disable tremolo
    ("v", 1): 0xC4,  #set volume
    ("v", 2): 0xC5,  #volume envelope
    ("v", 3): 0xCB,  #set tremolo
    ("&", 1): 0xE8,  #add to note duration
    ("<", 0): 0xD7,  #increment octave
    (">", 0): 0xD8,  #decrement octave
    ("[", 0): 0xE2,  #start loop
    ("[", 1): 0xE2,  #start loop
    ("]", 0): 0xE3  #end loop
    #(":", 1): 0xFC - jump to marker if event signal is sent
    #(";", 1): 0xF6 - jump to marker in another channel
    }

byte_tbl = {
    0xC4: (1, "v"),
    0xC5: (2, "v"),
    0xC6: (1, "p"),
    0xC7: (2, "p"),
    0xC8: (2, "m"),
    0xC9: (3, "m"),
    0xCA: (0, "m"),
    0xCB: (3, "v"),
    0xCC: (0, "v"),
    0xCD: (2, "p0,"),
    0xCE: (0, "p"),
    0xCF: (1, "%c"),
    0xD0: (0, "%n1"),
    0xD1: (0, "%n0"),
    0xD2: (0, "%p1"),
    0xD3: (0, "%p0"),
    0xD4: (0, "%e1"),
    0xD5: (0, "%e0"),
    0xD6: (1, "o"),
    0xD7: (0, "<"),
    0xD8: (0, ">"),
    0xD9: (1, "%k"),
    0xDA: (1, "m"),
    0xDB: (1, "k"),
    0xDC: (1, "@"),
    0xDD: (1, "%a"),
    0xDE: (1, "%y"),
    0xDF: (1, "%s"),
    0xE0: (1, "%r"),
    0xE1: (0, "%y"),
    0xE2: (1, "["),
    0xE3: (0, "]"),
    0xE4: (0, "%l1"),
    0xE5: (0, "%l0"),
    0xE6: (0, "%g1"),
    0xE7: (0, "%g0"),
    0xE8: (1, "&"),
    0xE9: (1, "s0"),
    0xEA: (1, "s1"),
    0xEB: (0, "\n;"),
    0xF0: (1, "t"),
    0xF1: (2, "t"),
    0xF2: (1, "%v"),
    0xF3: (2, "%v"),
    0xF4: (1, "%x"),
    0xF5: (3, "j"),
    0xF6: (2, "\n;"),
    0xF7: (2, "%b"),
    0xF8: (2, "%f"),
    0xF9: (0, "u1"),
    0xFA: (0, "u0"),
    0xFB: (0, '"'),
    0xFC: (2, ":"),
    0xFD: (1, "{FD}")
   }

equiv_tbl = {   #Commands that modify the same data, for state-aware modes (drums)
                #Treats k as the same command as v, though # params is not adjusted
    "v0,0": "v0",
    "p0,0": "p0",
    "v0,0,0": "v",
    "m0,0,0": "m",
    "|0": "@0",
    "%a": "%y",
    "%s": "%y",
    "%r": "%y",
    "|": "@0",
    "@": "@0",
    "o": "o0",
    }

mml2mfvi.py
import sys, os, re, traceback, copy
from string import maketrans
from mmltbl import *

mml_log = "\n" if __name__ == "__main__" else None

def byte_insert(data, position, newdata, maxlength=0, end=0):
    while position > len(data):
        data = data + "\x00"
    if end:
        maxlength = end - position + 1
    if maxlength and len(data) > maxlength:
        newdata = newdata[:maxlength]
    return data[:position] + newdata + data[position+len(newdata):]

    
def int_insert(data, position, newdata, length, reversed=True):
    n = int(newdata)
    l = []
    while len(l) < length:
        l.append(chr(n & 0xFF))
        n = n >> 8
    if not reversed: l.reverse()
    return byte_insert(data, position, "".join(l), length)

def warn(fileid, cmd, msg):
    global mml_log
    m = "{}: WARNING: in {:<10}: {}".format(fileid, cmd, msg)
    print m
    if __name__ == "__main__": mml_log += m + '\n'

def mlog(msg):
    global mml_log
    if __name__ == "__main__": mml_log += msg + '\n'
    
class Drum:
    def __init__(self, st):
        s = re.findall('(.)(.[+-]?)\\1=\s*([0-9]?)([a-gr^])([+-]?)\s*(.*)', st)
        if s: s = s[0]
        mlog("{} -> {}".format(st, s))
        if len(s) >= 6:
            self.delim = s[0]
            self.key = s[1]
            self.octave = int(s[2]) if s[2] else 5
            self.note = s[3] + s[4]
            s5 = re.sub('\s*', '', s[5]).lower()
            params = re.findall("\|[0-9a-f]|@0x[0-9a-f][0-9a-f]|%?[^|0-9][0-9,]*", s5)
            par = {}
            for p in params:
                if p[0] == "@" and len(p) >= 5:
                    if p[0:3] == "@0x":
                        par['@0'] = str(int(p[3:5], 16))
                        continue
                if p[0] == '|' and len(p) >= 2:
                    par['@0'] = str(int(p[1], 16) + 32)
                else:
                    pre = re.sub('[0-9]+', '0', p)
                    suf = re.sub('%?[^0-9]', '', p, 1)
                    if pre in equiv_tbl:
                        pre = equiv_tbl[pre]
                    par[pre] = suf
            self.params = par
        else:
            self.delim, self.key, self.octave, self.note, self.params = None, None, None, None, None
        mlog("DRUM: [{}] {} -- o{} {} {}".format(self.delim, self.key, self.octave, self.note, self.params))
        
def mml_to_akao(mml, fileid='mml', sfxmode=False, variant=None):
    #preprocessor
    #returns dict of (data, inst) tuples (4096, 32 bytes max)
    #one generated for each #VARIANT directive

    if isinstance(mml, str):
        mml = mml.splitlines()
    #one-to-one character replacement
    transes = []
    for line in mml:
        if line.startswith("#REPLACE") and len(line) > 7:
            tokens = line[7:].split()
            if len(tokens) < 3: continue
            if len(tokens[1]) != len(tokens[2]):
                warn(fileid, line, "token size mismatch, ignoring excess")
                if len(tokens[1]) > len(tokens[2]):
                    tokens[1] = tokens[1][0:len(tokens[2])]
                else:
                    tokens[2] = tokens[2][0:len(tokens[1])]
            transes.append(maketrans(tokens[1], tokens[2]))
    for trans in transes:
        newmml = []
        for line in mml:
            newmml.append(line.translate(trans))
        mml = newmml
    
    #sfxvariant
    all_delims = set()
    for line in mml:
        if line.startswith("#SFXV") and len(line) > 5:
            tokens = line[5:].split()
            if len(tokens) < 1: continue
            if len(tokens) >= 2 and sfxmode:
                all_delims.update(tokens[1])
            elif not sfxmode:
                all_delims.update(tokens[0])
    #variants
    variants = {}
    for line in mml:
        if line.startswith("#VARIANT") and len(line) > 8:
            makedefault = True if not variants else False
            tokens = line[8:].split()
            if len(tokens) < 1: continue
            if len(tokens) == 1:
                tokens.append('_default_')
            all_delims.update(tokens[0])
            variants[tokens[1]] = tokens[0]
            if makedefault: variants["_default_"] = tokens[0]
    for k, v in variants.items():
        variants[k] = "".join([c for c in all_delims if c not in variants[k]])
    if not variants:
        variants['_default_'] = ''.join([c for c in all_delims])
    if variant:
        if variant not in variants:
            print "mml error: requested unknown variant '{}'\n".format(variant)
        variants = {variant: variants[variant]}
        
    #generate instruments
    isets = {}
    for k, v in variants.items():
        iset = {}
        for line in mml:
            skip = False
            if line.startswith("#WAVE") and len(line) > 5:
                for c in v:
                    if c in line:
                        line = re.sub(c+'.*?'+c, '', line)
                line = re.sub('[^x\da-fA-F]', ' ', line[5:])
                tokens = line.split()
                if len(tokens) < 2: continue
                numbers = []
                for t in tokens[0:2]:
                    t = t.lower()
                    base = 16 if 'x' in t else 10
                    t = t.replace('x' if base == 16 else 'xabcdef', '')
                    try:
                        numbers.append(int(t, base))
                    except:
                        warn(fileid, "#WAVE {}, {}".format(tokens[0], tokens[1]), "Couldn't parse token {}".format(t))
                        continue
                if numbers[0] not in range(0x20,0x30):
                    warn(fileid, "#WAVE {}, {}".format(hex(numbers[0]), hex(numbers[1])), "Program ID out of range (expected 0x20 - 0x2F / 32 - 47)")
                    continue
                if numbers[1] not in range(0, 256):
                    warn(fileid, "#WAVE {}, {}".format(hex(numbers[0]), hex(numbers[1])), "Sample ID out of range (expected 0x00 - 0xFF / 0 - 255)")
                    continue
                iset[numbers[0]] = numbers[1]
        raw_iset = "\x00" * 0x20
        for slot, inst in iset.items():
            raw_iset = byte_insert(raw_iset, (slot - 0x20)*2, chr(inst))
        isets[k] = raw_iset
                
            
    #generate data
    datas = {}
    for k, v in variants.items():
        datas[k] = mml_to_akao_main(mml, v, fileid)
        
    output = {}
    for k, v in variants.items():
        output[k] = (datas[k], isets[k])
    
    return output
        
        
def mml_to_akao_main(mml, ignore='', fileid='mml'):
    mml = copy.copy(mml)
    #final bit of preprocessing
    macros = {}
    for line in mml:
        if line.lower().startswith("#def"):
            line = line[4:]
            line = line.split('#')[0].lower()
            if not line: continue
            pre, sep, post = line.partition('=')
            if post:
                pre = pre.replace("'", "").strip()
                for c in ignore:
                    try:
                        post = re.sub(c+".*?"+c, "", post)
                    except Exception:
                        c = "\\" + c
                        post = re.sub(c+".*?"+c, "", post)
                    post = "".join(post.split())
                macros[pre] = post.lower()
    
    stillmacros = True
    while stillmacros:
        stillmacros = False
        for i, line in enumerate(mml):
            for k, v in macros.items():
                if "'{}'".format(k) in line:
                    stillmacros = True
                    line = line.replace("'{}'".format(k), v)
            mml[i] = line
    #drums
    drums = {}
    for line in mml:
        if line.lower().startswith("#drum"):
            s = line[5:].strip()
            s = s.split('#')[0].lower()
            for c in ignore:
                s = re.sub(c+".*?"+c, "", s)
            for c in ["~", "/", "`", "\?", "_"]:
                s = re.sub(c, '', s)
            d = Drum(s.strip())
            if d.delim:
                if d.delim not in drums: drums[d.delim] = {}
                drums[d.delim][d.key] = d
    
    for i, line in enumerate(mml):
        mml[i] = line.split('#')[0].lower()
            
    m = list(" ".join(mml))
    targets, channels, pendingjumps = {}, {}, {}
    data = "\x00" * 0x26
    defaultlength = 8
    thissegment = 1
    next_jumpid = 1
    state = {}
    jumpout = []
    
    while len(m):
        command = m.pop(0)

        #conditionally executed statements
        if command in ignore:
            while len(m):
                next = m.pop(0)
                if next == command:
                    break
            continue
        #inline comment // channel marker
        elif command == "{":
            thisnumber = ""    
            numbers = []
            while len(m):
                command += m.pop(0)
                if command[-1] in "1234567890":
                    thisnumber += command[-1]
                elif thisnumber:
                    numbers.append(int(thisnumber))
                    thisnumber = ""
                if command[-1] == "}": break
            for n in numbers:
                if n <= 16 and n >= 1:
                    channels[n] = len(data)
            continue
        #drum mode
        elif command in drums:
            mls, dms = [], []
            drumset = drums[command]
            while len(m):
                if m[0] != command:
                    dms.append(m.pop(0))
                else:
                    m.pop(0)
                    break
            dbgdms = "".join(dms)
            lockstate = False
            if len(dms):
                if dms[0] in "1234567890":
                    state["o0"] = dms.pop(0)
                elif dms[0] == ">":
                    co = dms.pop(0)
                    while dms[0] == ">":
                        co += dms.pop(0)
                    state["o0"] += len(co)
                elif dms[0] == "<":
                    co = dms.pop(0)
                    while dms[0] == "<":
                        co += dms.pop(0)
                    state["o0"] -= len(co)
            while len(dms):
                dcom = dms.pop(0)
                if len(dms):
                    if dms[0] in "+-":
                        dcom += dms.pop(0)
                if dcom == "\\":
                    lockstate = True if not lockstate else False
                elif dcom == "!":
                    rcom = dms.pop(0)
                    if rcom == "!":
                        if "o0" in state:
                            state = {"o0": state["o0"]}
                        else:
                            state = {}
                        continue
                    if rcom == "%": rcom += dms.pop(0)
                    while len(dms):
                        if dms[0] in "0,":
                            rcom += dms.pop(0)
                        else: break
                    if rcom in equiv_tbl:
                        rcom = equiv_tbl[rcom]
                    state.pop(rcom, None)
                elif dcom in "0123456789^.":
                    mls.extend(dcom)
                elif dcom in drumset:
                    params = {}
                    for k, v in drumset[dcom].params.items():
                        if lockstate and k != "@0": continue
                        if k in state:
                            if state[k] != v:
                                params[k] = v
                        elif k == "%y" and not ( "%a0" in state or "%y0" in state or 
                                                 "%s0" in state or "%r0" in state):
                             pass
                        else:
                            params[k] = v
                    s = ""
                    for k, v in params.items():
                        t = (re.sub('[0-9,]', '', k) + v).strip()
                        s = t + s if k == "@0" else s + t
                        if k == "%y" or k == "@0":
                            state.pop("%a0", None)
                            state.pop("%y0", None)
                            state.pop("%s0", None)
                            state.pop("%r0", None)
                        if k != "%y": state[k] = v
                    if 'o0' in state:
                        if isinstance(state['o0'], str): state['o0'] = int(state['o0'])
                        ochg = drumset[dcom].octave - int(state['o0'])
                        if ochg < 0:
                            s += ">" * abs(ochg)
                        else: s += "<" * ochg
                        state['o0'] += ochg
                    else:
                        s += "o{}".format(drumset[dcom].octave)
                        state['o0'] = drumset[dcom].octave
                    s += drumset[dcom].note
                    mls.extend(list(s))
            mlog("drum: processed {} -> {}".format(dbgdms, "".join(mls)))
            mls.extend(m)
            m = mls
            continue
            
        #populate command variables
        if command == "%": command += m.pop(0)
        prefix = command
        if len(m):
            while m[0] in "1234567890,.+-x":
                command += m.pop(0)
                if not len(m): break
        
        #catch @0x before parsing params
        if "|" in command:
            command = "@0x2" + command[1:]
        if "@0x" in command:
            while len(command) < 5:
                command += m.pop(0)
            number = command[-2:]
            try:
                number = int(number, 16)
            except ValueError:
                warn(fileid, command, "Invalid instrument {}, falling back to 0x20".format(number))
                number = 0x20
            command = "@" + str(number)
                    
        modifier = ""
        params = []
        for c in command:
            if c in "+-":
                modifier = c
        thisnumber = ""
        for c in command[len(prefix):] + " ":
            if c in "1234567890":
                thisnumber += c
            elif thisnumber:
                params.append(int(thisnumber))
                thisnumber = ""
        dots = len([c for c in command if c == "."])
        
        if (prefix, len(params)) not in command_tbl and len(params):
            if (prefix + str(params[0]), len(params) - 1) in command_tbl:
                prefix += str(params.pop(0))
        
        #print "processing command {} -> {} {} mod {} dots {}".format(command, prefix, params, modifier, dots)
        #case: notes
        if prefix in "abcdefg^r":
            pitch = note_tbl[prefix]
            if prefix not in "^r":
                pitch += 1 if "+" in modifier else 0
                pitch -= 1 if "-" in modifier else 0
                while pitch < 0: pitch += 0xC
                while pitch > 0xB: pitch -= 0xC
            if not params:
                length = defaultlength
            else:
                length = params[0]
            if dots and str(length)+"." in length_tbl:
                akao = chr(pitch * 14 + length_tbl[str(length)+"."][0])
                dots -= 1
                length *= 2
            elif length in length_tbl:
                akao = chr(pitch * 14 + length_tbl[length][0])
            else:
                warn(fileid, command, "Unrecognized note length {}".format(length))
                continue
            while dots:
                if length*2 not in length_tbl:
                    warn(fileid, command, "Cannot extend note/tie of length {}".format(length))
                    break
                dots -= 1
                length *= 2
                if dots and str(length)+"." in length_tbl:
                    akao += chr(note_tbl["^"]*14 + length_tbl[str(length)+"."][0])
                    dots -= 1
                    length *= 2
                else:
                    akao += chr(note_tbl["^"]*14 + length_tbl[length][0])
            data += akao
        #case: simple commands
        elif (prefix, len(params)) in command_tbl:
            #special case: loops
            if prefix == "[":
                if len(params):
                    params[0] -= 1
                else:
                    params.append(1)
            #special case: portamento
            if prefix == "m" and len(params) == 2 and '-' in modifier:
                if params[1] > 0:
                    params[1] = 0x100 - params[1]
            #special case: end loop adds jump target if j,1 is used
            if prefix == "]":
                while len(jumpout):
                    pendingjumps[jumpout.pop()] = "jo%d"%next_jumpid
                targets["jo%d"%next_jumpid] = len(data) + 1
                next_jumpid += 1    
            #general case
            akao = chr(command_tbl[prefix, len(params)])
            #special case: pansweep
            if prefix == "p" and len(params) == 3:
                params = params[1:]
            #general case
            while len(params):
                if params[0] >= 256:
                    warn(fileid, command, "Parameter {} out of range, substituting 0".format(params[0]))
                    params[0] = 0
                akao += chr(params.pop(0))
            data += akao
        #case: default length
        elif prefix == "l" and len(params) == 1:
            if params[0] in length_tbl:
                defaultlength = params[0]
            else:
                warn(fileid, command, "Unrecognized note length {}".format(length))
        #case: jump point
        elif prefix == "$":
            if params:
                targets[params[0]] = len(data)
            else:
                targets["seg%d"%thissegment] = len(data)
        #case: end of segment
        elif prefix == ";":
            defaultlength = 8
            state = {}
            if params:
                if params[0] in targets:
                    target = targets[params[0]]
                else:
                    target = len(data)
                    pendingjumps[len(data)+1] = params[0]
            else:
                if "seg%d"%thissegment in targets:
                    target = targets["seg%d"%thissegment]
                else:
                    data += "\xEB"
                    thissegment += 1
                    continue
            data += "\xF6" + int_insert("  ",0,target,2)
            thissegment += 1
        #case: jump out of loop
        elif prefix == "j":
            if len(params) == 1:
                jumpout.append(len(data)+2)
                target = len(data)
            elif len(params) == 2:
                if params[1] in targets:
                    target = targets[params[1]]
                else:
                    target = len(data)
                    pendingjumps[len(data)+2] = params[1]
            else: continue
            if params[0] >= 256:
                warn(fileid, command, "Parameter {} out of range, substituting 1".format(params[0]))
                params[0] = 1
            data += "\xF5" + chr(params[0]) + int_insert("  ",0,target,2)
        #case: conditional jump
        elif prefix == ":" and len(params) == 1:
            if params[0] in targets:
                target = targets[params[0]]
            else:
                target = len(data)
                pendingjumps[len(data)+1] = params[0]
            data += "\xFC" + int_insert("  ",0,target,2)
    
    #insert pending jumps
    for k, v in pendingjumps.items():
        if v in targets:
            data = int_insert(data, k, targets[v], 2)
        else:
            warn(fileid, command, "Jump destination {} not found in file".format(v))
    #set up header
    header = int_insert("\x00"*0x26, 0, len(data)-2, 2)
    header = int_insert(header, 2, 0x26, 2)
    header = int_insert(header, 4, len(data), 2)
    for i in xrange(0,8):
        if i not in channels:
            channels[i] = len(data)
    for k, v in channels.items():
        header = int_insert(header, 4 + k*2, v, 2)
        if k <= 8 and k+8 not in channels:
            header = int_insert(header, 4 + (k+8)*2, v, 2)
    data = byte_insert(data, 0, header, 0x26)
    
    return data
    
def clean_end():
    print "Processing ended."
    raw_input("Press enter to close.")
    quit()
    
if __name__ == "__main__":
    mml_log = "\n"

    print "mfvitools MML to AKAO SNESv4 converter"
    print
    
    if len(sys.argv) >= 2:
        fn = sys.argv[1]
    else:
        print "Enter MML filename.."
        fn = raw_input(" > ").replace('"','').strip()
    
    try:
        with open(fn, 'r') as f:
            mml = f.readlines()
    except IOError:
        print "Error reading file {}".format(fn)
        clean_end()

    try:
        variants = mml_to_akao(mml)
    except Exception:
        traceback.print_exc()
        clean_end()
    
    fn = os.path.splitext(fn)[0]
    for k, v in variants.items():
        vfn = ".bin" if k in ("_default_", "") else "_{}.bin".format(k)
        
        thisfn = fn + "_data" + vfn
        try:
            with open(thisfn, 'wb') as f:
                f.write(v[0])
        except IOError:
            print "Error writing file {}".format(thisfn)
            clean_end()
        print "Wrote {} - {} bytes".format(thisfn, hex(len(v[0])))
        
        thisfn = fn + "_inst" + vfn
        try:
            with open(thisfn, 'wb') as f:
                f.write(v[1])
        except IOError:
            print "Error writing file {}".format(thisfn)
            clean_end()
        print "Wrote {}".format(thisfn)
        
    try:
        with open(os.path.join(os.path.split(sys.argv[0])[0],"mml_log.txt"), 'w') as f:
            f.write(mml_log)
    except IOError:
        print "Couldn't write log file, displaying..."
        print mml_log
            
    print "Conversion successful."
    print
    
    clean_end()
